remove-user-on-logout/src/main/java/de/intension/listener/RemoveUserOnLogOutEventListenerProvider.java

```diff
@@ -1,68 +1,72 @@
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.models.UserProvider;
 
 /**
  * Event listener to remove user on logout for the users from identity providers.
  */
 public class RemoveUserOnLogOutEventListenerProvider
-    implements EventListenerProvider
-{
+        implements EventListenerProvider {
 
-    private static final Logger   LOG = Logger.getLogger(RemoveUserOnLogOutEventListenerProvider.class);
+    private static final Logger LOG = Logger.getLogger(RemoveUserOnLogOutEventListenerProvider.class);
 
     private final KeycloakSession keycloakSession;
 
-    protected RemoveUserOnLogOutEventListenerProvider(KeycloakSession session)
-    {
+    private final EventListenerTransaction tx = new EventListenerTransaction(null, this::removeUser);
+
+    protected RemoveUserOnLogOutEventListenerProvider(KeycloakSession session) {
         this.keycloakSession = session;
+        session.getTransactionManager().enlistAfterCompletion(tx);
         LOG.debugf("[%s] instantiated.", this.getClass());
     }
 
     @Override
-    public void onEvent(Event event)
-    {
+    public void onEvent(Event event) {
         EventType eventType = event.getType();
 
         if (EventType.LOGOUT.equals(eventType)) {
-            RealmModel realm = keycloakSession.getContext().getRealm();
-            String userId = event.getUserId();
-            UserProvider provider = this.keycloakSession.getProvider(UserProvider.class, "jpa");
-            UserModel userModel = provider.getUserById(realm, userId);
-            Stream<FederatedIdentityModel> federatedIdentity = keycloakSession.users()
-                .getFederatedIdentitiesStream(realm, userModel);
-            if (federatedIdentity.findAny().isPresent()){
-                removeUser(realm, userModel, userId);
-            }
+            tx.addEvent(event);
         }
     }
 
     /**
-     * Remove user by userId and cancel related 'remove expired session users' task.
+     * Remove user from keycloak on logout.
+     * Starts a jpa transaction before removing the user since eventlistener tarnsaction does not hava a running jpa-transaction
      */
-    private void removeUser(RealmModel realm, UserModel userModel, String userId)
-    {
-        keycloakSession.users().removeUser(realm, userModel);
-        keycloakSession.getTransactionManager().commit();
-        LOG.debugf("User with id [%s] is removed on logout.", userId);
+    private void removeUser(Event event) {
+        EntityManager entityManager = keycloakSession.getProvider(JpaConnectionProvider.class).getEntityManager();
+        JpaKeycloakTransaction transaction = new JpaKeycloakTransaction(entityManager);
+        transaction.begin();
+        RealmModel realm = keycloakSession.getContext().getRealm();
+        UserProvider userProvider = keycloakSession.users();
+
+        UserModel user = userProvider.getUserById(realm, event.getUserId());
+        if (user != null) {
```

Comment:
is this requirement changed now to remove all the users without checking the IDP link?
Or just decided to do so as this is gonna be only in VIDIS realm?

Reply:
Discussed this this week with Florian and Markus because there are a lot of user on the system that do not have a link to their idp (anymore). They do have a user attribute though but snce they said that no users in vidis-realm are needed, I didn't want to slow down the deletion by adding this check.